zebedee-cms/src/test/java/com/github/onsdigital/zebedee/keyring/LegacyKeyringImpl_PopulateTest.java

```diff
@@ -0,0 +1,209 @@
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, null, null));
+
+        assertThat(ex.getMessage(), equalTo(EMAIL_EMPTY_ERR));
+    }
+
+    @Test
+    public void testPopulate_srcUserEmailEmpty_ShouldThrowException() {
+        when(srcUser.getEmail())
+                .thenReturn("");
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, null, null));
+
+        assertThat(ex.getMessage(), equalTo(EMAIL_EMPTY_ERR));
+    }
+
+    @Test
+    public void testPopulate_srcUserKeyringNull_ShouldThrowException() {
+        when(srcUser.keyring())
+                .thenReturn(null);
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, null, null));
+
+        assertThat(ex.getMessage(), equalTo(USER_KEYRING_NULL_ERR));
+    }
+
+    @Test
+    public void testPopulate_srcUserKeyringNotInCache_ShouldThrowException() throws Exception {
+        when(keyringCache.get(srcUser))
+                .thenReturn(null);
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, null, null));
+
+        assertThat(ex.getMessage(), equalTo(CACHE_KEYRING_NULL_ERR));
+        verify(keyringCache, times(1)).get(srcUser);
+    }
+
+    @Test
+    public void testPopulate_srcKeyringLocked_ShouldThrowException() throws Exception {
+        when(srcCachedKeyring.isUnlocked())
+                .thenReturn(false);
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, null, null));
+
+        assertThat(ex.getMessage(), equalTo(KEYRING_LOCKED_ERR));
+        verify(keyringCache, times(1)).get(srcUser);
+    }
+
+    @Test
+    public void testPopulate_targetUserNull_ShouldThrowException() {
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, null, null));
+
+        assertThat(ex.getMessage(), equalTo(USER_NULL_ERR));
+    }
+
+    @Test
+    public void testPopulate_targetUserEmailNull_ShouldThrowException() {
+        when(targeUser.getEmail())
+                .thenReturn(null);
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, targeUser, null));
+
+        assertThat(ex.getMessage(), equalTo(EMAIL_EMPTY_ERR));
+    }
+
+    @Test
+    public void testPopulate_targetUserEmailEmpty_ShouldThrowException() {
+        when(targeUser.getEmail())
+                .thenReturn("");
+
+        KeyringException ex = assertThrows(KeyringException.class,
+                () -> legacyKeyring.populate(srcUser, targeUser, null));
+
+        assertThat(ex.getMessage(), equalTo(EMAIL_EMPTY_ERR));
+    }
+
+    @Test
+    public void testPopulate_targetUserKeyringNull_ShouldThrowException() {
```

Comment:
To begin with the target user Keyring is null should we not return a new HashSet?

Reply:
We could but the expectation is this should not be null. If it is null something has gone wrong and we should throw an error to draw attention to this. As you've suggested we could "recover" this situation by simply creating a new map but this would be hiding the fact something was not right/not as expected. 

Others might disagree with me on this by my philosophy is it's almost always better to error when something it not as expected and draw attention to it. 

When that happens we can investigate then and work out if its OK or if there really is an issue and take the appropriate action.
If we hide that something has gone wrong it could cause other problems in different parts of the app or manifest itself in subtle ways which can cause confusion or weird behaviour.